### DISCUSSION TITLE
What do I do about a bug that can't include a "code file"?

### DISCUSSION THREAD

**Discussion by brandonzylstra** [ddnexus/pagy#729] [Q&A] [answered]
I just found a bug in the Pagy gem, but it's not a bug that appears in my app, it's a bug that appears when using the `pagy` executable according to [the directions found in the documentation](https://ddnexus.github.io/pagy/playground/#3-demo-app).

Here's the text I would be posting in the bug report, if not for the threat of automatically closing bug reports that don't have a code file (thus eliminating the possibility of reporting bugs for which no code file _can_ be included, and thus guaranteeing that bugs of this sort are never reported or the reports are never seen):

---

The documentation claims that you can see the demos without Bundler by mentioning Bundler as something you might _possibly_ need to use.  Yet when I try to see them without Bundler, I am met with errors:

```shell
brandon@Air ~ % gem update pagy
Updating installed gems
Nothing to update
Gems already up-to-date: pagy
brandon@Air ~ % pagy calendar
/Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:328:in `rescue in root': Could not locate Gemfile or .bundle/ directory (Bundler::GemfileNotFound)
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:324:in `root'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:101:in `bundle_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:621:in `configure_gem_home_and_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:87:in `configure'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/gems/3.3.0/gems/pagy-9.0.4/bin/pagy:52:in `<top (required)>'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `load'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `<main>'
/Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler/shared_helpers.rb:17:in `root': Could not locate Gemfile (Bundler::GemfileNotFound)
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:325:in `root'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:101:in `bundle_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:621:in `configure_gem_home_and_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:87:in `configure'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/gems/3.3.0/gems/pagy-9.0.4/bin/pagy:52:in `<top (required)>'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `load'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `<main>'
brandon@Air ~ % pagy demo
/Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:328:in `rescue in root': Could not locate Gemfile or .bundle/ directory (Bundler::GemfileNotFound)
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:324:in `root'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:101:in `bundle_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:621:in `configure_gem_home_and_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:87:in `configure'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/gems/3.3.0/gems/pagy-9.0.4/bin/pagy:52:in `<top (required)>'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `load'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `<main>'
/Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler/shared_helpers.rb:17:in `root': Could not locate Gemfile (Bundler::GemfileNotFound)
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:325:in `root'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:101:in `bundle_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:621:in `configure_gem_home_and_path'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/site_ruby/3.3.0/bundler.rb:87:in `configure'
	from /Users/brandon/.rbenv/versions/3.3.4/lib/ruby/gems/3.3.0/gems/pagy-9.0.4/bin/pagy:52:in `<top (required)>'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `load'
	from /Users/brandon/.rbenv/versions/3.3.4/bin/pagy:25:in `<main>'	
```

If the documentation is in error, then this could be reclassified as a request to correct the docs, but I'm assuming the docs reflect the intended behavior of the gem, and this is merely an oversight because most of us are using a Gemfile almost all of the time.

**Comment by ddnexus** [comment]
The docs could obviously be improved, sice you obviously didn't get it.

What it means is that depending on your environment, it could work with or without the bundler. Now... you got the error without bundler, so that means that in your case you need `bundle exec`.

I actually don't get your complaint, but if you have some better sentence to use, we are open to suggestions.

**Reply by brandonzylstra** [reply]
> The docs could obviously be improved, sice you obviously didn't get it.

@ddnexus What do I "obviously" not get?

> What it means is that depending on your environment, it could work with or without the bundler. Now... you got the error without bundler, so that means that in your case you need `bundle exec`.

I installed the gem with `gem install`.  Adding `bundle exec` would be useless to find a gem not in a Gemfile.  **_The problem is that the code is failing.**  Perhaps when Pagy was originally written it supported the use case of installing without a Gemfile, and that use case was neglected as the code has been changed.  

> I actually don't get your complaint, but if you have some better sentence to use, we are open to suggestions.

Unless your intent is for a Gemfile to be necessary--which would mean your intent changed since the docs were written--then the problem is that the demos did not work.  There's a bug in the code, but it's not in my app since I'm not in this instance using Pagy in any app, so I can't send you any code file.  That means that in addition to the bug in Pagy you also have a problem with your standards for bug submission that make it impossible to submit a report for such a bug.

**Comment by ddnexus** [comment]
> What do I "obviously" not get?

I thought you didn't get the meaning of the sentence, hence I thought that the sentence may not be clear and I asked if you have a better way of saying it. It was self criticism about the docs, not about your understanding.

> I installed the gem with `gem install`. Adding `bundle exec` would be useless to find a gem not in a Gemfile. 

"useless to find a gem not in a Gemfile"... of course. But that is not the only thing `bundle exec` does.

> The problem is that the code is failing.

No, here it's not failing at all. What is failing is your assumption that the doc is wrong because you don't use bundle to install the pagy gem and think that it would be useless. But the playground apps use bundle inline to install the gems that they use. Depending on your environment you can call `pagy` without the `bundle exec`, still having the apps using `bundle inline`... exactly like the doc says.

> Unless your intent is for a Gemfile to be necessary--which would mean your intent changed since the docs were written

Never changed

> then the problem is that the demos did not work.

The demo does work here... just follow the docs

> There's a bug in the code,

No bug as long as you use the launch statement that applies AFAIK.

> That means that in addition to the bug in Pagy you also have a problem with your standards for bug submission that make it impossible to submit a report for such a bug.

A lot of wrong assumptions here. 

If there is a bug you CAN reproduce it. For example, just posting the app that show the failure of some pagination feature even without modification. That would be the easiest reproducible bug. Of course if you cannot run the app because you want to use the wrong statement for your environment, it's a problem of environment (and wrong assumptions), not an app bug and much less a pagy bug.

Indeed if you find a bug in the playground system, that wouldn't even be a gem core bug. That would affect the playground, not the pagination, however you could absolutely use the app to show the app bug.

HTH

**Reply by ddnexus** [reply]
BTW, since you didn't suggest anything... would just using `bundle exec`, dropping the case without it, be a better alternative? It's probably more familiar for everybody, and for the cases that work without it, it wouldn't be much more effort.

**Reply by ddnexus** [reply]
We spent too much time on this... I just pushed the change to the docs. HTH

**Comment by ddnexus** [comment, marked as answer]
Hey @brandondrew,

you were totally right about being a bug!!!

I just thought I checked it in isolation, but obviously I did not. My env was masking the problem, and I said a bunch of stupid things. Sorry for that.

I had a `Bundle.configure` in the mix (useful when we use `--no-install` working on the project), not protected by any condition, and that was causing the problem.

The upcoming pagy v10 is going to fix it.
Thanks and Cheers

**Reply by brandonzylstra** [reply]
Thank you for the apology and admission of error.

**Reply by ddnexus** [reply]
Thank you for the report and understanding!

**Reply by brandonzylstra** [reply]
You're _very_ welcome for the report.  You've created a remarkable gem that makes everyone's life easier, and submitting an error report is a minuscule contribution--negligible in comparison.

The understanding...well, that's a little more difficult to be frank...

**Reply by ddnexus** [reply]
> The understanding...well, that's a little more difficult to be frank...

Yeah, I get it! it's difficult to believe that I am just a human 👽 🤣🤣🤣
